Add a "Set as Featured Image" Button to Image Block

SiobhyB opened the pull request:
With this PR, we'll be implementing a "set as featured image" button with the purpose of making it easier for users to assign a new featured image directly from the post area.

Fixes https://github.com/wordpress-mobile/gutenberg-mobile/issues/1011

`gutenberg-mobile`: wordpress-mobile/gutenberg-mobile#3116
`gutenberg`: https://github.com/WordPress/gutenberg/pull/28854

Please refer to the Gutenberg PR as the central place for steps to tests and screenshots.

PR submission checklist:

- [x] I have considered adding unit tests where possible.
- [x] I have considered adding accessibility improvements for my changes.
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

SiobhyB committed:
Update GutenbergEditorFragment.java

This commit is mostly for testing purposes, I've updated GutenbergEditorFragment.java with a function called "onSetFeaturedImageButtonClicked" and want to test the results.

SiobhyB committed:
Update reference to latest gutenberg-mobile commit

This commit updated reference to latest gutenberg-mobile commit.

SiobhyB committed:
Set featured image

This commit introduces code that sets a featured image when the "set as featured" button is clicked from an image block. A function within "EditPostSettingsFragment" is currently utilised for this purpose, though this is likely not the optimal approach.

SiobhyB committed:
Bump Gutenberg mobile

SiobhyB committed:
Updates to GutenbergEditorFragment

This commit cleans up some of the changes that have been made to the GutenbergEditorFragment file, deleting unused/commented out code, etc.

SiobhyB committed:
Tidy up EditPostActivity

SiobhyB committed:
Tidy up EditorFragmentAbstract

SiobhyB committed:
Tidy up EditorFragmentAbstract

Fix some issues with spacing

SiobhyB committed:
Fix accidental deletion of "final" from expression

SiobhyB committed:
Restore spacing in GutenbergEditorFragment

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB commented:
The PR is still in draft mode because, though the code technically works to set a featured image, the approach I’ve taken feels fragile/error-prone. 

I’ve added a new `updateFeaturedImage` method under [the _EditorFragmentListener methods_ section](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2915) in **EditPostsActivity**. That method utilizes the existing `updateFeaturedImage` function [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java#L1001) in **EditPostSettingsFragment**. I then call this function in **GutenbergEditorFragment**.

By calling the function in this way, our data tools won't distinguish between any clicks to the image block's new featured image button and the button within the Post Settings screen. It also feels fragile, like any issues with **EditPostSettingsFragment** may then lead to issues with the image block's button.   

I'm looking to explore other patterns/ways to set the featured image from **GutenbergEditorFragment**. 

cc-ing @hypest (who noted he'd be able to take a look over this later in the week) and @jd-alexander (who has familiarity with the code from buddying me with it these past weeks). Thanks in advance for your thoughts and any pointers!

hypest commented:
Looks like the bundling job fails with `error index.js: Cannot find module 'babel-plugin-transform-remove-console'` and that's the error I also see locally while trying to build WPAndroid on this PR. Will try to understand what's going on there.

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB committed:
Bump gutenberg-mobile

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -795,6 +808,39 @@ public void onClick(DialogInterface dialog, int id) {
+    private void featuredImageSet(int mediaId) {
+        mEditorFragmentListener.updateFeaturedImage(mediaId, true);
+        setFeaturedImageId(mediaId);
+        ToastUtils.showToast(getActivity(), R.string.featured_image_confirmation).show();
```
👋 Siobhan, I'm assuming that the Toast here is mostly for debugging purposes and not part of the final design, right? If so, what do you think about adding a "TODO" comment here to help us remember to remove it before we finalize the feature?

SiobhyB replied:
You're right, I'll add a "TODO" and this is a great lesson for me to be more diligent in doing similar in the future.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:
```diff
@@ -2974,6 +2974,11 @@ public void onEditPostPublishedSettingsClick() {
 
+    @Override
+    public void updateFeaturedImage(final long mediaId, final boolean imagePicked) {
+        mEditPostSettingsFragment.updateFeaturedImage(mediaId, imagePicked);
```
Any particular reason to not reuse the [already defined `setFeaturedImageId`](https://github.com/wordpress-mobile/WordPress-Android/blob/05ddeced286ff7ab0ee67cfe856b178f0af6f0d7/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java#L2550)? That one calls the same function in `mEditPostSettingsFragment` and also checks if the reference is null, making it slightly more safe.

SiobhyB replied:
Ah, only my lack of knowledge. :D I hadn't paid much attention to that function within EditPostActivity, I'd been too focused on inspecting the functions under EditPostSettings. I've updated this now and it works great.

hypest commented on libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java:
```diff
@@ -795,6 +808,39 @@ public void onClick(DialogInterface dialog, int id) {
+    }
+
+    private void featuredImageSet(int mediaId) {
+        mEditorFragmentListener.updateFeaturedImage(mediaId, true);
```
Looking at [the code that uses the `imagePicked` parameter](https://github.com/wordpress-mobile/WordPress-Android/blob/9f9a9c9d3b024debc62bf5b3211691d6390b1059/WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostSettingsFragment.java#L1002-L1008) (passed as `true` here), my impression is that it uses that param just to trigger a Tracks event during the media picking flow.

During the set-feature-image flow from inside an image block, I'd expect to pass `false` so to avoid triggering that event so, I'm not sure why the `true` here @SiobhyB , can you elaborate?

This also seems related to the concern you expressed in [the comment](https://github.com/wordpress-mobile/WordPress-Android/pull/14052#issuecomment-801029547).

hypest commented:
> It also feels fragile, like any issues with EditPostSettingsFragment may then lead to issues with the image block's button.

Can you elaborate on that @SiobhyB ? Like, what could be an example of an issue with EditPostSettingsFragment that'd break the image block button?

In the meantime, in case that's what you have in mind too, it _is_ a bit awkward that the editor fragment needs to call a function in the Post Settings fragment just to update the featured image id. Feels like such a function could live in a headless "service" fragment inside the EditPostActivity instead of lying inside a UI fragment.

I still think that the current PR is a good first take on implementing the set-featured-image functionality, and a cleaner approach (perhaps by extracting the shared code that sets the featured image id in the local DB) could be worked on in a follow up PR.

SiobhyB committed:
Bump gutenberg-mobile

SiobhyB committed:
Introduce "OnSetFeaturedImageListener"

This commit introduces an "OnSetFeaturedImageListener" and moves new featured image related functions to it from "OnMediaLibraryButtonListener".

SiobhyB committed:
Added TODO to clarify Toast isn't part of final design

SiobhyB committed:
Add Snackbar and bump Gutenberg Mobile

The main purpose of this commit is to replcace the current Toast confirmation when a featured image is set with a Snackbar.

SiobhyB committed:
Bump

SiobhyB committed:
Fixing merge conflict

SiobhyB commented:
Thank you for looking into this @hypest! 🙇‍♀️ 

> Can you elaborate on that @SiobhyB ? Like, what could be an example of an issue with EditPostSettingsFragment that'd break the image block button?

This was more of a _feeling_ than anything concrete I felt certain about. Your note about it being a bit awkward to call the function in the Post Settings fragment matches up to the hunch I had and I wasn't sure if I was missing wider problems this could bring due to my lack of experience.

> In the meantime, in case that's what you have in mind too, it is a bit awkward that the editor fragment needs to call a function in the Post Settings fragment just to update the featured image id. Feels like such a function could live in a headless "service" fragment inside the EditPostActivity instead of lying inside a UI fragment.

Great, thank you! If there aren't any major issues then I'll work with this approach for now. :)

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB committed:
Update image blocks when featured image's are set in Post Settings

SiobhyB committed:
Introduce "removeFeaturedImage" function and update name of "onGetFeaturedImageId"

This commit introduces a "removeFeaturedImage" function to allow users to remove a featured image directly from the image block. It also updates the name of the "onGetFeaturedImageId" to "checkIfFeaturedImage" in order to better reflect its current purpose.

SiobhyB committed:
Update function name from "onRequestFeaturedImageId" to "sendToJSFeaturedImageId"

This commit updates "onRequestFeaturedImageId" to "sendToJSFeaturedImageId", in order to clarify the purpose of the function.

SiobhyB committed:
Remove SnackBar for removed featured image

The SnackBar is currently necessary for when a featured image is removed, so removing for now.

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB committed:
Add Tracks events for new featured image buttons

This commit enables tracking for the image block's two new featured image buttons ("set as featured image" and "remove as featured image").

hypest commented:
👋 @SiobhyB , looks like this PR is not up-to-date with the latest from the gutenberg-mobile/Gutenebrg side PRs and won't compile correctly. Can you update it? While at it, I wonder if it might be a better strategy to test/develop by using the WPAndroid app anyway (instead of the gb-mobile demo apps), since the feature includes many changes in the bridge anyway.

SiobhyB committed:
Bump Gutenberg-Mobile

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB commented:
I've updated this PR now, thanks again for looking!

> While at it, I wonder if it might be a better strategy to test/develop by using the WPAndroid app anyway (instead of the gb-mobile demo apps), since the feature includes many changes in the bridge anyway.

I've been developing/testing using the WPAndroid app in order to allow me to test the changes I'm making. I'm still actively working on and committing to this issue while it's in draft mode, which I think is why you caught the PR while it's out-of-date. Let me know if I'm missing something else here.

_Edited to add: I'm aware that there are merge conflicts here, it's on my list to look into!_

SiobhyB committed:
Bump Gutenberg Mobile

SiobhyB committed:
Merge branch 'develop' into update/image-block-with-featured-image-setting

SiobhyB committed:
Address merge conflicts

SiobhyB committed:
Remove Snackbar

The native Snackbar is going to be replaced by the new Compact Notice on the JS side.

SiobhyB committed:
Remove Snackbar reference and bump Gutenberg Mobile

SiobhyB committed:
"sendToJSFeaturedImageId" name changes and addition of notices

This commits tidies up some functions that had not yet been renamed to "sendToJSFeaturedImageId", and also sends notices to the JS side when images are set or removed as featured.

SiobhyB committed:
Bump Mobile Gutenberg

SiobhyB committed:
Introduce "featuredMedia" prop to "GutenbergPropsBuilder"

SiobhyB committed:
Remove checkIfFeaturedImage() function

SiobhyB committed:
Remove logging from EditPostActivity

This commit removes logging I'd placed in the EditPostActivity file for debugging purposes.

SiobhyB committed:
Update gutenberg mobile reference

SiobhyB committed:
Merge branch 'develop' of github.com:wordpress-mobile/WordPress-Android into update/image-block-with-featured-image-setting

hypest commented:
Just a quick note: we weren't exactly trying to merge this PR but got merged automatically because we merged [a PR that wholly included](https://github.com/wordpress-mobile/WordPress-Android/pull/14451) the commits already.
